File: leetcode/java/37.sudoku-solver.java
class Solution {
    public void solveSudoku(char[][] board) {
        backtrack(board, 0, 0);
    }

    boolean validNumber(char[][] board, int i, int j, char c) {
        // row
        for (int k = 0; k < 9; k++) {
            if (c == board[i][k]) {
                return false;
            }
        }
        // col
        for (int k = 0; k < 9; k++) {
            if (c == board[k][j]) {
                return false;
            }
        }
        // 3*3
        int row = i / 3;
        int col = j / 3;
        for (int k = row * 3; k < (row + 1) * 3; k++) {
            for (int l = col * 3; l < (col + 1) * 3; l++) {
                if (board[k][l] == c) {
                    return false;
                }
            }
        }
        return true;
    }

    boolean backtrack(char[][] board, int i, int j) {
        if (j == 9) {
            i++;
            j = 0;

            if (i == 9) {
                return true;
            }
        }
        if (board[i][j] != '.') {
            return backtrack(board, i, j + 1);
        }

        for (int k = 1; k < 10; k++) {

            if (validNumber(board, i, j, (char) ('0' + k))) {
                board[i][j] = (char) ('0' + k);
                boolean res = backtrack(board, i, j + 1);
                if (res) {
                    return true;
                } else {
                    board[i][j] = '.';
                }
            }
        }
        return false;
    }
}

File: leetcode/python/37.sudoku-solver.py
from typing import List


class Solution:
    def solveSudoku(self, board: List[List[str]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """

        def can_place(row, col, s: str) -> bool:
            r = row // 3
            c = col // 3
            for i in range(len(board)):
                if board[row][i] == s:
                    return False
                if board[i][col] == s:
                    return False
            for i in range(3):
                for j in range(3):
                    if board[r * 3 + i][c * 3 + j] == s:
                        return False
            return True

        def back_track(row, col):

            if col == 9:
                row += 1
                col = 0
            if row == 9:
                return True
            if board[row][col] != '.':
                return back_track(row, col + 1)

            for i in range(1, 10):
                if can_place(row, col, str(i)):
                    board[row][col] = str(i)

                    if back_track(row, col + 1):
                        return True
                    board[row][col] = '.'

            return False

        back_track(0, 0)
